src/topology/schema.rs

```diff
@@ -30,82 +31,59 @@ use crate::{
 ///
 /// Finally, The merged definition (named `Definition 1 & 2`), and `Definition 4` are merged
 /// together to produce the new `Definition` returned by this method.
-pub(super) fn merged_definition(
+pub fn merged_definition(
     inputs: &[OutputId],
     config: &dyn ComponentContainer,
-    cache: &mut HashMap<Vec<OutputId>, Definition>,
+    cache: &mut HashMap<(bool, Vec<OutputId>), Definition>,
 ) -> Definition {
     // Try to get the definition from the cache.
-    if let Some(definition) = cache.get(inputs) {
+    if let Some(definition) = cache.get(&(config.schema_enabled(), inputs.to_vec())) {
         return definition.clone();
     }
 
-    // This should be "never", but that isn't supported yet (coming soon!)
-    let mut definition: Option<Definition> = None;
+    // TODO: special case when inputs is empty?
+    let mut definition = Definition::empty_kind(Kind::never(), []);
 
     for input in inputs {
         let key = &input.component;
 
-        // If the input is a source, it'll always have schema definition attached, even if it is an
-        // "empty" schema.
-        //
-        // We merge this schema into the top-level schema.
```

Comment:
There are some code comments that were added for clarity to future readers that got removed in some of these changes. I'd like to see those added back (rewritten where needed).

Reply:
Many of the comments were just repeating verbatim what the code is doing. For example

> // For sources, a `None` schema definition is equal to an "empty" definition.

That's what the code is saying already, so it's not really adding anything valuable.
```unwrap_or_else(Definition::empty),```

Same thing for this one

```
// If we find no match, it means the topology is misconfigured. This is a fatal
// error, but other parts of the topology builder deal with this state.
None => unreachable!("source output misconfigured"),
```
The `unreachable` already means it's a fatal error and the code should be preventing this state. The message passed in already explains why it could be reached. The comment is just duplicating this.

In some cases I think the comments were here because the code was a bit hard to read. For example, the `outputs.iter().find_map` usage was more difficult to read, and was refactored to simplify it.

Reply:
Just provoding another point of reference here: Even if the comment repeats what the code is doing, that helps (me) a lot when unpacking compact logic. Most of the comments seemed to also provide additional context that might be missed when just looking at the local code

Reply:
Given Pablo's comment, I'm also inclined to ask you to document this one again, either using existing documentation (if applicable) or some new comments, just so we don't lose track of how all of this is supposed to work, in the future.

Reply:
I added more comments. Let me know if there's anything else here potentially confusing / needs more.